pkg/apis/messaging/v1alpha1/choice_types.go

```diff
@@ -0,0 +1,138 @@
+// Choice defines conditional branches that will be wired in
+// series through Channels and Subscriptions.
+type Choice struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Choice.
+	Spec ChoiceSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Choice. This data may be out of
+	// date.
+	// +optional
+	Status ChoiceStatus `json:"status,omitempty"`
+}
+
+// Check that Sequence can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Choice)(nil)
+var _ apis.Defaultable = (*Choice)(nil)
+
+// TODO: make appropriate fields immutable.
+//var _ apis.Immutable = (*Choice)(nil)
+var _ runtime.Object = (*Choice)(nil)
+var _ webhook.GenericCRD = (*Choice)(nil)
+
+type ChoiceSpec struct {
+	// Cases is the list of Filter/Subscribers pairs. Filters are evaluated in the order
+	// provided, until one pass (returns true)
+	Cases []ChoiceCase `json:"cases"`
+
+	// ChannelTemplate specifies which Channel CRD to use
+	ChannelTemplate ChannelTemplateSpec `json:"channelTemplate"`
+
+	// Reply is a Reference to where the result of the last Subscriber of a given branch gets sent to.
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	//
+	//  The resource pointed by this ObjectReference must meet the Addressable contract
+	//  with a reference to the Addressable duck type. If the resource does not meet this contract,
+	//  it will be reflected in the Subscription's status.
+	// +optional
+	Reply *corev1.ObjectReference `json:"reply,omitempty"`
+}
+
+type ChoiceCase struct {
+	// Filter is the expression guarding the branch/case
+	Filter *eventingv1alpha1.SubscriberSpec `json:"filter,omitempty"`
+
+	// Subscriber receiving the event when the filter passes
+	Subscriber eventingv1alpha1.SubscriberSpec `json:"subscriber"`
+}
+
+type ChoiceChannelStatus struct {
+	// Channel is the reference to the underlying channel.
+	Channel corev1.ObjectReference `json:"channel"`
+
+	// ReadyCondition indicates whether the Channel is ready or not.
+	ReadyCondition apis.Condition `json:"ready"`
+}
+
+type ChoiceSubscriptionStatus struct {
+	// Subscription is the reference to the underlying Subscription.
+	Subscription corev1.ObjectReference `json:"subscription"`
+
+	// ReadyCondition indicates whether the Subscription is ready or not.
+	ReadyCondition apis.Condition `json:"ready"`
+}
+
+// ChoiceStatus represents the current state of a Choice.
+type ChoiceStatus struct {
+	// inherits duck/v1alpha1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1beta1.Status `json:",inline"`
+
+	// ChoiceStatuses is an array of corresponding Subscription statuses.
```

Comment:
Wonder if it would be better to have a ChoiseCaseStatus and then an array here. I think it's little more complicated than just subscriptionstatus/channelstatus, because each Case actually has 2 subscriptions? It has one that subscribes the filter to the "ingress" channel, and then there's a subscription created from filter (and I'd assume a channel between the filter / "actual" subscriber), so at the very least it seems like you'd need to represent state of 2 subscriptions in each CaseStatus as well as probably one channel (between filter / "actual" subscription) and maybe a Channel after the "actual" subscriber that a Reply gets subscribed to?

Reply:
yup agree. The flat array was just to get started.

Reply:
We might also want to propagate the Reply status to make it clear which one has been applied (global reply or case reply).